James Webadmin CLI help command

quantranhong1999 opened the pull request:
write Picocli help command and its unit test, packaged it with GraalVM native image maven plugin.
have 5 Checkstyle violations, love to be reviewed.

chibenwa commented on pom.xml:
```diff
@@ -540,6 +540,7 @@
         <module>server</module>
         <module>testing/base</module>
         <module>third-party</module>
+        <module>server/protocols/webadmin-cli</module>
```
This module have to be added in server/pom.xml (and respect alphabetic ordering)

chibenwa commented on server/protocols/webadmin-cli/pom.xml:
```diff
@@ -0,0 +1,65 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
You miss the compulsary license

chibenwa commented on server/protocols/webadmin-cli/pom.xml:
```diff
@@ -0,0 +1,65 @@
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <parent>
+        <artifactId>james-project</artifactId>
+        <groupId>org.apache.james</groupId>
```
server project should be the parent

chibenwa commented on server/protocols/webadmin-cli/pom.xml:
```diff
@@ -0,0 +1,65 @@
+    </parent>
+    <modelVersion>4.0.0</modelVersion>
+
+    <artifactId>webadmin-cli</artifactId>
```
Add a name and a description tag

chibenwa commented on server/protocols/webadmin-cli/pom.xml:
```diff
@@ -0,0 +1,65 @@
+        </plugins>
+    </build>
+
+    <dependencies>
```
dependencies go first, then build

chibenwa commented on server/protocols/webadmin-cli/pom.xml:
```diff
@@ -0,0 +1,65 @@
+            <version>20.2.0</version>
+            <scope>provided</scope>
+        </dependency>
+        <!-- https://mvnrepository.com/artifact/org.junit.jupiter/junit-jupiter-api -->
```
These comments are not needed

chibenwa commented on server/protocols/webadmin-cli/pom.xml:
```diff
@@ -0,0 +1,65 @@
+    <properties>
+        <maven.compiler.source>11</maven.compiler.source>
+        <maven.compiler.target>11</maven.compiler.target>
+    </properties>
```
put properties at the top, before dependencies

chibenwa commented on server/protocols/webadmin-cli/src/main/java/org/apache/james/Picocli/Picocli.java:
```diff
@@ -0,0 +1,23 @@
+package org.apache.james.Picocli;
```
you miss the license

chibenwa commented on server/protocols/webadmin-cli/src/main/java/org/apache/james/Picocli/Picocli.java:
```diff
@@ -0,0 +1,23 @@
+    @Override
+    public void run() {
+        final String helpMessage = "There is no help for any commands";
+        if (isNeedHelp) System.out.println(helpMessage);
```
You likely don't pass the checkstyle :-)

Try to run `mvn clean package -DskipTests` on your work

chibenwa replied:
We should not generate our own help message but let picocli generate it.

See https://picocli.info/#execute

chibenwa commented on server/protocols/webadmin-cli/src/main/java/org/apache/james/WebAdminCli.java:
```diff
@@ -0,0 +1,16 @@
+package org.apache.james;
```
License... (configure your IDE)

chibenwa commented on server/protocols/webadmin-cli/src/test/java/UnitTest/HelpCommandTest.java:
```diff
@@ -0,0 +1,35 @@
+package UnitTest;
```
License

chibenwa commented on server/protocols/webadmin-cli/src/test/java/UnitTest/HelpCommandTest.java:
```diff
@@ -0,0 +1,35 @@
+    @Test
+    void allHelpCommandShouldWork(){
+        new CommandLine(new Picocli()).execute("--help");
+        Assertions.assertEquals(successHelpMessage, outputStreamCaptor.toString().trim());
```
Use assertj library, set testing base as a test dependency

chibenwa commented on server/protocols/webadmin-cli/pom.xml:
```diff
@@ -0,0 +1,65 @@
+        </dependency>
+        <!-- https://mvnrepository.com/artifact/org.junit.jupiter/junit-jupiter-api -->
+        <dependency>
+            <groupId>org.junit.jupiter</groupId>
```
do not depend directly on junit, set testing base as a test dependency

chibenwa reviewed (commented):
Good work so far!

Please remove the readme and adr from this pr...

Arsnael commented on server/protocols/webadmin-cli/src/test/java/UnitTest/HelpCommandTest.java:
```diff
@@ -0,0 +1,35 @@
+    private final ByteArrayOutputStream outputStreamCaptor = new ByteArrayOutputStream();
+
+    @BeforeEach
+    void setup(){
```
missing space between `()` and `{`
Same comment for below methods

Arsnael reviewed (commented):
Nice work overall :)

chibenwa commented on server/protocols/webadmin-cli/pom.xml:
```diff
@@ -1,17 +1,66 @@
+    <properties>
+        <maven.compiler.source>11</maven.compiler.source>
+        <maven.compiler.target>11</maven.compiler.target>
+    </properties>
```
Are these properties needed?

chibenwa commented on server/protocols/webadmin-cli/src/main/java/org/apache/james/Picocli/Picocli.java:
```diff
@@ -1,23 +1,37 @@
+ * under the License.                                             *
+ ******************************************************************/
+
 package org.apache.james.Picocli;
```
I propose org.apache.james.cli as a package

Picocli is an implementation detail. Also don't use capital letters in packages, reserve that for classes.

chibenwa commented on server/protocols/webadmin-cli/src/main/java/org/apache/james/Picocli/Picocli.java:
```diff
@@ -1,23 +1,37 @@
-            description = "Display help for commands"
-    )
-    public boolean isNeedHelp;
+public class Picocli implements Runnable {
```
Let's delete this Picocli class and move everything back to WebAdminCli ?

chibenwa commented on server/protocols/webadmin-cli/src/test/java/org/apache/james/Picocli/HelpVersionCommandTest.java:
```diff
@@ -0,0 +1,78 @@
+
+public class HelpVersionCommandTest {
+
+    private static final String successHelpMessage = "Usage: ./james-cli [-hV]\n" +
```
Use capital snake case for constants: `SUCCESS_HELP_MESSAGE`

chibenwa commented on server/protocols/webadmin-cli/src/test/java/org/apache/james/Picocli/HelpVersionCommandTest.java:
```diff
@@ -0,0 +1,78 @@
+            "  -h, --help      Show this help message and exit.\n" +
+            "  -V, --version   Print version information and exit.";
+
+    private static final String rightVersion = "1.0";
```
`VERSION`

chibenwa commented on server/protocols/webadmin-cli/src/test/java/org/apache/james/Picocli/HelpVersionCommandTest.java:
```diff
@@ -0,0 +1,78 @@
+import java.io.ByteArrayOutputStream;
+import java.io.PrintStream;
+
+import static org.assertj.core.api.Assertions.*;
```
We don't use wildard import. Import everything separately.

You might need to configure your IDE regading this.

Don't hesitate to document this in the env setup page.

chibenwa commented on server/protocols/webadmin-cli/src/main/java/org/apache/james/WebAdminCli.java:
```diff
@@ -19,9 +19,14 @@
 
 package org.apache.james;
```
`package org.apache.james.cli;`

chibenwa commented on server/protocols/webadmin-cli/pom.xml:
```diff
@@ -56,6 +56,17 @@
+        </dependency>
+        <dependency>
+            <groupId>io.github.openfeign</groupId>
+            <artifactId>feign-gson</artifactId>
```
Feign is not needed yet right?

chibenwa commented on server/protocols/webadmin-cli/src/main/java/org/apache/james/cli/WebAdminCli.java:
```diff
@@ -17,18 +17,38 @@
-package org.apache.james;
+package org.apache.james.cli;
 
+import org.apache.james.cli.domain.DomainListCommand;
```
What is that DomainListCommand doing in a "help message PR ?" :-)

quantranhong1999 commented:
Sorry i messed this up :( . This PR help command branch, i removed domain part by hand. But the domain list command branch, it's broken. I think because of I dropped some commit (related to pom.xml, WebAdminCli.java, HelpVersionCommandTest ) while i rebase.
I would like to take some advice how to avoid this situation. Is there a way that i could remove the domain list part from this PR not by hand?

quantranhong1999 commented:
Do I need to be accepted this help command PR before i open PR for domain list command?
If not, there will be help command files change(vs linagora:master) in that domain list command PR. How can I remove them from the domain list PR ?

chibenwa commented:
You can start working on domain list.

We generally put unrelated dependency changes (here help) in a commit prefixed by `[NO REVIEW]`

Also please run mvn package on this pr: i suspect compile or checkstyle issues...

quantranhong1999 commented:
> You can start working on domain list.
> We generally put unrelated dependency changes (here help) in a commit prefixed by `[NO REVIEW]`
> Also please run mvn package on this pr: i suspect compile or checkstyle issues...

I run mvn clean package for thís PR => Build success, no more checkstyle violations
About unrelated dependency changes, what about files changes part? Do we need to get rid of them from new PR?

chibenwa commented:
```
[dede32125ac1d98c684939e728a6f12f36c0932a] [ERROR] Failed to execute goal org.graalvm.nativeimage:native-image-maven-plugin:20.2.0:native-image (default) on project webadmin-cli: Execution default of goal org.graalvm.nativeimage:native-image-maven-plugin:20.2.0:native-image failed: A required class was missing while executing org.graalvm.nativeimage:native-image-maven-plugin:20.2.0:native-image: jdk/vm/ci/services/Services
```

chibenwa commented:
> About unrelated dependency changes, what about files changes part? Do we need to get rid of them from new PR?

This PR looks good.

Other PRs can include commits of this work - with a `[NO_REVIEW]` prefix.

quantranhong1999 commented:
> ```
> [dede32125ac1d98c684939e728a6f12f36c0932a] [ERROR] Failed to execute goal org.graalvm.nativeimage:native-image-maven-plugin:20.2.0:native-image (default) on project webadmin-cli: Execution default of goal org.graalvm.nativeimage:native-image-maven-plugin:20.2.0:native-image failed: A required class was missing while executing org.graalvm.nativeimage:native-image-maven-plugin:20.2.0:native-image: jdk/vm/ci/services/Services
> ```

This plugin requires download graalvm community version  and set it as JAVA_HOME to build native image.
https://www.graalvm.org/docs/getting-started-with-graalvm/linux/
I will write a graalvm environment document if you want.

chibenwa commented:
>This plugin requires download graalvm community version and set it as JAVA_HOME to build native image.
> https://www.graalvm.org/docs/getting-started-with-graalvm/linux/
> I will write a graalvm environment document if you want.

No we should ideally not require extra setup.

chibenwa commented:
If it is not possible to do that, as we won't require extra build dependencies, that would be a no-go for GraalVM, and we should be using JAT packaging.

chibenwa commented:
I started a discussion about that around https://issues.apache.org/jira/browse/JAMES-3400

chibenwa commented:
Following JIRA comments, we won't use graalvm but instead maven-jar plugin.

Can you do the switch here, in the README of the CLI, and also modify the ADR ?

chibenwa reviewed (commented):
Modify the README to adapt it to JAR packaging. Idem for the ADR.

quantranhong1999 commented:
> Modify the README to adapt it to JAR packaging. Idem for the ADR.

I opened a separately PR for ADR and README

Arsnael commented on server/protocols/webadmin-cli/pom.xml:
```diff
@@ -0,0 +1,88 @@
+            <version>4.5.1</version>
+        </dependency>
+        <dependency>
+            <groupId>${james.groupId}</groupId>
```
The dependencies are ordered by alphabetical order, on the groupId then artifactId.

Except for the james dependencies `<groupId>${james.groupId}</groupId>`, thos always go before third party dependencies

Arsnael reviewed (approved):
Just a little extra comment, but I approve overall :)

chibenwa commented:
test this please

build output lost

chibenwa commented:
test this please

quantranhong1999 commented:
> test this please

I built webadmin-cli successful.
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  19.141 s
[INFO] Finished at: 2020-10-23T09:11:31+07:00
[INFO] ------------------------------------------------------------------------

> build output lost
Can you tell me more details?

chibenwa commented on server/protocols/webadmin-cli/src/test/java/org/apache/james/cli/HelpVersionCommandTest.java:
```diff
@@ -49,13 +49,15 @@ void tearDown() {
 
     @Test
     void shortHelpCommandShouldShowHelpMessage() {
         new CommandLine(new WebAdminCli()).execute("-h");
```
`WebAdminCli.main("-h")` ?

Or at least `WebAdminCli.executeFluent("-h")`

quantranhong1999 commented:
1 Checkstyle violation

chibenwa commented:
```
[fc3469210e3d32bfcea766c08adff6e5a26865f2] [ERROR] Failed to execute goal on project james-server-util: Could not resolve dependencies for project org.apache.james:james-server-util:jar:3.6.0-SNAPSHOT: Failed to collect dependencies at com.github.steveash.guavate:guavate:jar:1.0.0 -> com.google.guava:guava:jar:25.1-jre -> org.codehaus.mojo:animal-sniffer-annotations:jar:1.14: Failed to read artifact descriptor for org.codehaus.mojo:animal-sniffer-annotations:jar:1.14: Could not transfer artifact org.codehaus.mojo:animal-sniffer-annotations:pom:1.14 from/to central (https://repo.maven.apache.org/maven2): /root/.m2/repository/org/codehaus/mojo/animal-sniffer-annotations/1.14/animal-sniffer-annotations-1.14.pom.part (No such file or directory) -> [Help 1]
```

test this please

chibenwa commented:
`[ERROR] error reading /root/.m2/repository/com/github/fge/throwing-lambdas/0.5.0/throwing-lambdas-0.5.0.jar; zip file is empty`

test this please

Arsnael commented:
```
[fc3469210e3d32bfcea766c08adff6e5a26865f2] [INFO] There is 1 error reported by Checkstyle 8.29 with checkstyle.xml ruleset.
[fc3469210e3d32bfcea766c08adff6e5a26865f2] [ERROR] src/main/java/org/apache/james/cli/WebAdminCli.java:[24] (imports) ImportOrder: Wrong order for 'java.util.concurrent.Callable' import.
```

chibenwa commented:
> Ask for someone to help you configure the imports on your IDE

@Arsnael  can you do it?

chibenwa commented:
Merged
